### DISCUSSION TITLE
Curve fitting

### DISCUSSION THREAD

**Discussion by fire** [iimachines/Maya2glTF#158] [General]
Heya, I was wondering how the reducer branch works, I also need an arbitrary transforms reducer for Godot Engine.

**Comment by ziriax** [comment]
These were just some experiments to see what kind of compression could be achieved with very straightforward schemes.

I found that for my particular use case, chebyshev polynomials performed really well. But I didn't have time to experiment with radial basis functions, or advanced techniques like motion matching.

I suggest to just use the Animation Compression Library:

https://github.com/nfrechette/acl

It has experimental glTF support

https://github.com/nfrechette/acl-gltf

**Reply by fire** [reply]
I found that I wanted curve fitting and ACL when I checked last a year ago uses bitwiddling which I wanted to throw ZSTD at.

**Reply by ziriax** [reply]
My customer didn’t want any decompression, arguing that this would be too slow (not my personal opinion).

So I made a scheme where the curve is fitted brute force with chebyshev polynomials. These polynomials have the amazing property that you can truncate them and still get a very good approximation. So basically I ended up with a very simple array with offsets to these coefficients. Overall I was pleased with this: it is simple, compresses well, and decompresses on the fly.

However in the end my customer went for fully procedural IK driven animation, so the compression experiment was abandoned.

**Reply by fire** [reply]
Is there a good alternative for alglib.spline1dbuildakima in c++?

**Comment by ziriax** [comment]
Sorry, I don't know. I can only ask Google:

https://github.com/tzolkincz/fast-akima-interpolation

**Reply by fire** [reply]
I decided to use https://www.geometrictools.com/Source/Interpolation.html instead of https://www.boost.org/doc/libs/develop/libs/math/doc/html/math_toolkit/makima.html

**Reply by fire** [reply]
Can you describe how you go from akima to chebyshev polynomials. 

Also how do you treat orientation differently than vectors?

**Comment by ziriax** [comment]
This has been a while, but as far as I know you can directly apply the
Chevy transform to the samples, no need for any intermediate interpolation
spline.
